lib/dao.js

```diff
@@ -624,6 +624,101 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert(arguments.length >= 1, 'At least one argument is required');
+  if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
+    data = where;
+    where = {};
+  } else if (options === undefined && cb === undefined) {
+    if (typeof data === 'function') {
+      cb = data;
+      data = where;
+      where = {};
+    }
+  } else if (cb === undefined) {
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+  data = data || {};
+  options = options || {};
+  cb = cb || utils.createPromiseCallback();
+  assert(typeof where === 'object', 'The where argument must be an object');
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+  var hookState = {};
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+  var modelObject = connector._models;
+  var modelName = Object.keys(modelObject)[0];
+  if (connector.upsertWithWhere) {
+    return connector.upsertWithWhere(modelName, where, data, cb);
+  } else {
+    var element = Object.keys(where)[0];
+    var filter = { 'where': where };
+    if (Object.keys(data).length != 0) {
+      if (!data.hasOwnProperty(element)) {
+        data[element] = where[element];
+      }
+      var pk = idName(Model);
+      self.find(filter, function(err, instances) {
+        var modelsLength = instances.length;
+        if (modelsLength === 0 && data.hasOwnProperty(pk)) {
+          return self.create(data, options, cb);
+        } else if (modelsLength == 0 && data.hasOwnProperty(pk) == false) {
+          process.nextTick(function() {
+            cb('Error message: No matching instance found.Found error while trying ' +
+              'to create a new instance.' +
+            'Primary key' +
+            'is not provided in the payload');
+          });
+        } else if (modelsLength === 1) {
+          var pkName = idName(Model);
+          var idVal = instances[0][pkName];
+          self.findById(idVal, function(err, modelInst) {
+            if (err) {
+              process.nextTick(function() {
```

Comment:
ditto. Both `find` and `findById` already defers callback to the next tick, no need to call `process.nextTick` here.